# RxJava 异常处理

## RxJavaPlugins.setErrorHandler 

我们可以在`Application`中设置这个全局的`ErrorHandler`:

```
    RxJavaPlugins.setErrorHandler {

    }
```

这个回调会在以下时机调用:

1. `subscribe`时的`onError`中代码发生异常时
2. `subscribe`时没有设置`onError`的Consumer,并且`onNext`中发生异常时

上面这两个逻辑可以在`RxJava`的源码中看出, 以我们在`subscribe`是直接写`lambda表达式`为例:

```
public final class LambdaObserver<T> ...{

    @Override
    public void onNext(T t) {
        if (!isDisposed()) {
            try {
                onNext.accept(t);
            } catch (Throwable e) {
                Exceptions.throwIfFatal(e);
                get().dispose();
                onError(e); // onNext发生异常回调onError
            }
        }
    }

    @Override
    public void onError(Throwable t) {
        if (!isDisposed()) {
            lazySet(DisposableHelper.DISPOSED);
            try {
                onError.accept(t); //如果subscribe时没有设置onError,则这里的onError为null
            } catch (Throwable e) {
                Exceptions.throwIfFatal(e);
                RxJavaPlugins.onError(new CompositeException(t, e));  //onError发生异常时回调全局的设置
            }
        } else {
            RxJavaPlugins.onError(t);
        }
    }
}
```

要知道的是:**如果设置了`RxJavaPlugins.setErrorHandler`,那么在发生异常时应用可能根本没有任何表现**。在开发时其实这是一个比较蛋疼的事，因为你不知道哪里出错了。。

并且`RxJavaPlugins.setErrorHandler`回调中的`Exception`并不是我们崩溃的`Exception`。它的作用主要像一个兜底的异常处理者。

# App崩溃日志

一般app都会接入像`bugly`这样的崩溃分析平台。当应用发生崩溃时在`bugly`上就可以找到。不过我们在开发时如果应用发生崩溃，这时候去`bugly`上找这个异常可能还是有点麻烦的。

我们可以使用`Java UncaughtExceptionHandler`来捕获应用的崩溃，然后做一些处理。比如保存日志到本地来方便我们查看具体的崩溃信息。

## UncaughtExceptionHandler

我们可以使用它来处理`Java`非检查异常。它可以设置给某个线程，也可以设置给全部的线程:

```
Thread.setDefaultUncaughtExceptionHandler(new UncaughtExceptionHandler() {
    @Override
    public void uncaughtException(Thread t, Throwable e) {
        System.err.println("程序抛出了一个异常，异常类型为 ： " + e);
    }
});
```

上面给全部的线程设置了`UncaughtExceptionHandler`,即如果线程发生崩溃，就会回调`UncaughtExceptionHandler.uncaughtException()`。

设置`UncaughtExceptionHandler`的另外一个目的是捕获**子线程中的异常**, 要知道我们再主线程中是捕获不到子线程的异常的哦。

>`kotlin`中没有受检的异常 : http://www.kotlincn.net/docs/reference/exceptions.html


# 抓取主线程异常






